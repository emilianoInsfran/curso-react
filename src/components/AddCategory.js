import React, { useState } from 'react';
import PropTypes from 'prop-types';//validaciones

export const AddCategory = ( { setCategory } )=> {//utilice la desestructuraciÃ³n de objetos, no use los props (quedaria sino props.setCategory)

    const [inputValue, setInputValue] = useState('Hola mundo');

    const handleInputChange = (e)=>{
        setInputValue(e.target.value);
        console.log("Se llamo handleInputChange");
    }

    const handleSubmit = (e) =>{
        e.preventDefault()

        console.log("se llamo a handlesubmit",inputValue);
        if(inputValue.trim().length > 2){//trim borra los espacio dentro de un string
            setCategory(cats => [ inputValue,...cats,])
            setInputValue('');
        }
    }

    return (
        <>
            <form onSubmit= { handleSubmit }>
                <input
                    type="text"
                    value={ inputValue }
                    onChange={ handleInputChange }
                /> 
            </form>
        </>
    )
}

AddCategory.propTypes = {
    setCategory: PropTypes.func.isRequired,//le digo que tieene que ser una funcion y que sea obligatorio (props) - 
}