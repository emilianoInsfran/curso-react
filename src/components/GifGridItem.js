import React from 'react';
import PropTypes from 'prop-types';

export const GifGridItem = ( {id,title,url} ) => {
    // //como class es una palabra reservada de js le pusieron className
    return (
        <div className="card">
            <img src={url} title={title}/>
            <h2>{title}</h2>
        </div>
    )
} 


GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}