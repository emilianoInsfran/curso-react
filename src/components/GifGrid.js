import React, { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
//useEffect: me parmite ejecutar sierto codigo de manera condicional

export const GifGrid = ({ category }) => {

    const  [count, setCount] = useState(0); 
    //const [ images, setImages] = useState([]); 

    // useEffect( () => {//quiero que se ejecute cuando el componente es renderizado por primera vez
    //     getGifs(category).then(setImages);
    // },[category])//el segundo argumento [] es una array de argumentos si quiero que se ejecute una vez lo dejo como un array vacio, pero si la categoria cambia se vuelve a ejecutar
//{loading && <p>Loading</p>} si es verdadero pone el loading, sino, no hace nada. otra forma de validaci√≥n
//podemos renombrar data  poniendo dos puntos: data:images
    const {data:images,loading} = useFetchGifs(category);//custom hooks
    return (
        <>
        <div>
            <h3>{ category }</h3>
            <h3>{ loading? 'cargando...' : 'data cargada' }</h3>
            {loading && <p>Loading</p>}
            <button onClick={()=>setCount( count + 1 )}>+1</button>

            <ol>
                {
                    images.map( (img) =>(
                    // images.map( ({id,title}) =>(
                        <GifGridItem key={img.id} {...img}/>//estoy enviando cada propiedad como una independiente (es un ejemplo pero usas los props)
                        // <li key={id}>{ title }</li>
                    ))
                }
            </ol>
        </div>
        </>
    )
}