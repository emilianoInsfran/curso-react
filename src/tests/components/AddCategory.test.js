import React from 'react';
import { shallow } from 'enzyme';
import { AddCategory } from '../../components/AddCategory';
import '@testing-library/jest-dom';


describe('Pruebas con getGif Fetch',()=>{
    const setCategory = jest.fn();
    let wrapper = shallow( <AddCategory setCategory= { setCategory } /> );

    beforeEach(()=>{
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategory= { setCategory } /> );
    });


    test('debe de mostrarse correctamente ', ()=>{
        expect( wrapper ).toMatchSnapshot();
    });

    test('debe cambiar la caja de texto ', ()=>{
        const input = wrapper.find('input');
        const value = 'Hola mundo';
        input.simulate('change',{
            target: {
                value:value
            }
        });
    });

    test('No deve mostrar la informaciÃ³n si se hizo submit',()=>{
        wrapper.find('form').simulate('submit',{preventDefault(){}});
        expect(setCategory).toHaveBeenCalled();
    })


});