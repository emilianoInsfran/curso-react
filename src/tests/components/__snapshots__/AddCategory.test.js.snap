// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con getGif Fetch debe de mostrarse correctamente  1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value="Hola mundo"
    />
  </form>
</Fragment>
`;
